### ISSUE TITLE
Regression in the XProc harness for Saxon

### ISSUE THREAD

**Issue by fgeorges** [xspec/xspec#97]
The XProc harness for Saxon (and apparently the others as well) seem to have suffered a regression.  They are requiring `xspec-home` to be set regardless of the way they are used, even though the reference to the sources where all carefully checked (resolving using `xspec-home` if set, or using the global name URI if not).

It seems now they are requiring `xspec-home` anyway.  Not sure what happened.

**Comment by cirulls** [comment]
Hi Florent, 

we noticed that the XProc harness was broken as reported in  #74 and #72. We fixed this in #76 by reverting the XProc to the version currently integrated with Oxygen and added a test to run the XProc harness with Saxon in order to catch any future regression bugs (see [xspec.bats](https://github.com/xspec/xspec/blob/master/test/xspec.bats#L195-199)). This implementation requires ```xspec-home``` to be set as it was in v0.3.0 integrated in Oxygen. 

These changes were required in order to have a working and usable XProc harness for Saxon. There is now documentation in the [wiki](https://github.com/xspec/xspec/blob/master/test/xspec.bats#L195-199) as well.

I hope this clarifies the issue. 

Sandro

**Comment by fgeorges** [comment]
Well, I am not sure.  #74 looks like `xspec-home` was not defined in that case.  It looked like a matter of setting it somehow, rather than reverting months of work?

I am still not sure about why it has been reverted to an older version rather than fixing the bug itself.

**Comment by AirQuick** [comment]
As discussed in #53, the master branch requires more stability than the past trunk. For a couple of particular cases, the long standing v0.3 final seemed more stable than the intermediate v0.4 RC.
That's why I agreed to reverting, with this [comment](https://github.com/xspec/xspec/issues/74#issuecomment-278097670): "until someone gets interested seriously in improving the multiple processors support".
I think you're the very person. Can you please help resume improving XProc harnesses?

**Comment by fgeorges** [comment]
Not sure what part of #53 you refer to.  Still not seeing the rationale here.  I had a bit of time, but I have to admit this complicates everything.

**Comment by fgeorges** [comment]
I have put back the newest version of the harnesses in https://github.com/xspec/xspec/tree/fix/revert-reverting.  I fixed the bug with the input port not bound on the error step, and the xspec-home parameter not used in the format step.

Can you validate it works for you, and solve the problem you had initially (from your ticket description, it should), so I can merge it to master?

**Comment by AirQuick** [comment]
All I can say at this point is that the master branch's status badge is now red...

**Comment by cirulls** [comment]
Sorry for my delayed reply, I was busy at work today and didn't want to reply with a quick comment as this needs some clarification. I'm replying here to #99 and #100 as well. This is going to be a long reply so please bear with me.

As @AirQuick stated in https://github.com/xspec/xspec/issues/99#issuecomment-284736326, the rationale for reverting the XProc harness to v0.3.0 and removing attributes ```assert``` and ```context``` was to have:

- working features that did not break other parts of the codebase
- code that is tested and documented

The XProc harness was not working as it was implemented in v0.4.0-RC as reported in #74 and #72. The introduction of attributes ```assert``` and ```context``` in v0.4.0-RC raised issues #26, #30, #31, #32, and #33. We could not find these attributes in the schema [xspec.rnc](https://github.com/xspec/xspec/blob/master/src/schemas/xspec.rnc) nor documentation. This does not mean that these features are not useful, for example I can definitely see the need for an assert statement and leaner XProc harness suites. However, the people who were involved in these issues (this includes me) were not able to find a quick fix or to figure out certain parts of the code base. 

As @AirQuick already said, this is where we need your help Florent. You have a great deal of experience with XSpec and other XML technologies and were behind some of the design decisions that increased the adoption of XSpec. We would all love to get your pull requests and get an experienced developer like you for XSpec. There are however some basic rules we all need to follow in order to make XSpec thrive:

- Decisions on the code base are taken by consensus as often as possible and by majority in worst case scenarios. I want to make clear that XSpec is not my personal pet project. I may not contribute to this project in few years' time and I want to see it growing even when key maintainers leave. I personally believe this is the very nature of open source projects.
- On a technical level, this means that every change in the code base is done via pull requests and the pull request is reviewed and approved by someone else before squash merging it into master. We informally followed this model until now and this has greatly increased the overall code quality (this was certainly the case for pull requests initiated by me and reviewed by @AirQuick). This rule applied to everybody and especially to users like me, @AirQuick, @fgeorges, and @tgraham-antenna who have write permissions on the XSpec repository. 
- New features and bugs fixes should be provided with unit tests before merging them into master. Tests can also be written by other people or, if a test cannot be provided, there should be a very good reason (e.g. we had tests that required Saxon EE which we cannot currently run on Travis and AppVeyor). Tests prevent regressions and act as self-documentation which then allow to write proper documentation in the wiki for new features. 
- We cannot merge into master until the test suite running on Travis and AppVeyor is green. The test suite runs automatically on Travis for Unix systems and AppVeyor for Windows every time there is commit, a pull request or a merge in the repository. This scenario should normally not occur if we follow the procedure above. There are other projects that rely on a stable XSpec repository (I once broke myself @tofi86's builds in #19 and was made aware of the importance of a green badge). 

I'm happy to discuss this further if people feel the need or disagree with the basic rules I listed above. I don't think GitHub is a good place for this sort of conversation and I would prefer to organise a Skype call or something along those lines. I hope we can find a consensus and I really don't want to get this nasty as I'm spending my limited spare time in this project. Plus, I prefer to spend this time writing code rather than lenghty comments like this one. 

As the master branch is currently red, I'm going to raise a pull request to revert it to the previous commit and ask @fgeorges and @AirQuick to review it. This does not mean that the change is not needed, the rationale is simply to have a green test suite on master branch until the code changes pass all the tests. Again, this is to avoid creating issues to other projects relying on the XSpec repository.  

Florent, as I said above, we really look forward to your pull requests and really would like to get your experience and contributions for XSpec.

**Comment by fgeorges** [comment]
Well, replying the same day is not really "delayed answer", is it? :-)

Unfortunately, your questions require to invest some time in it, which I cannot provide in the short time future.  I am spending as much spare time I got (actually more than that) in getting the master green again.

**Comment by AirQuick** [comment]
"resolving using `xspec-home` if set, or using the global name URI if not)" exists in v2.0.7:

https://github.com/xspec/xspec/blob/552300776faa1c245d89a1b2ead7d1a83443bf60/src/harnesses/harness-lib.xpl#L125-L133
https://github.com/xspec/xspec/blob/552300776faa1c245d89a1b2ead7d1a83443bf60/src/harnesses/harness-lib.xpl#L192-L200
https://github.com/xspec/xspec/blob/552300776faa1c245d89a1b2ead7d1a83443bf60/src/harnesses/harness-lib.xpl#L269-L274
